_posts/2017/2017-04-02-rest-collections-hal-and-http2.md

```diff
@@ -0,0 +1,424 @@
+            },
+
+            "title" : "..",
+            "pubDate" : "..",
+            "content" : ".."
+        }
+     ]
+
+  }
+}
+```
+
+Two things happened here. Each individual item in the collection now appears
+in `_embedded`. They have their own sets of links, including a `self` link,
+referring to the uri of the resource we just embedded. We also removed the
+items from `_links`.
+
+The way we look at `_embedded` is that things that appear in `_embedded` are:
+
+* Link relations, equal to items appearing in `_links`.
+* The data you _would_ receive, if you did a `GET` request on the target of
+  the link.
+
+This is important, because a good HAL-based REST client should ideally
+consider things appearing in `_embedded` and `_links` as the exact same thing.
+The only difference is that because an item appears in `_embedded`, it's no
+longer needed to perform that `GET` request. The client should cache it.
+
+If a client is built around this core concept, another benefit is that the
+client becomes adaptable to changes to the server. You might for example see
+over time that a HAL client very often might follow a certain link and almost
+always will want the data for it.
+
+
+### An example
+
+Here's a real-world example from our API. Our API has a document on the root
+of the API that the client uses to discover all the other resources. It
+contains a link to a resource that has information about the current user:
+
+```json
+{
+  "_links" : {
+     "self" : { "href" : "/" },
+     "current-user" : { "href" : "/user/1356" },
+     "support" : { "href": "mailto:support@example.org" }
+  }
+}
+```
+
+We noticed that all clients _always_ requests the current users' information
+after logging in. It always follows up this initial `GET` to a `GET` to
+whatever the `current-user` relation points at.
+
+Knowing this, we can adapt our API to simply assume this and
+pre-emptively send that information over:
+
+
+```json
+{
+  "_links" : {
+     "self" : { "href" : "/" },
+     "support" : { "href": "mailto:support@example.org" }
+  },
+  "_embedded" : {
+     "current-user" : {
+        "_links" : {
+            "self" : { "href" : "/user/1356" }
+        },
+        "firstName" : "...",
+        "lastName" : "...",
+        "email" : "..."
+     }
+  }
+}
+```
+
+A good HAL client would need 0 changes, and simply adapt to this new situation
+and skip the second GET request.
+
+
+A few problems with this
+------------------------
+
+One of the biggest advantages and promises that REST typically gives us, is
+by using the functionality HTTP brings out of the box, we get all the benefits
+that HTTP brings itself. One of the biggest examples that is always given, is
+the rich caching features HTTP has.
+
+However, HTTP caches will not be aware of embedded resources. In the last
+example, the cache doesn't know that `/user/1356` was embedded and cached, and
+it doesn't know it does not know it can skip a future `GET` request to that
```

Comment:
Nah I still appreciate. I just haven't really read the whole thing front to back myself yet, so it just made me feel bad for not doing better spell checking and stuff